### ISSUE TITLE
Memory corruption when using mpfr_get_str and mpfr_free_str?

### ISSUE THREAD

**Issue by SanderBouwhuis** [BrianGladman/mpfr#17]
I'm running into trouble when using **mpfr_get_str** and **mpfr_free_str**. It underruns the buffer and thereby corrupts the memory.

I have the following code:
```
// This function determines the actual significant decimals needed to display the value
uint32 CBigFloat::GetTotSignificantDecimals() const
{
  uint32 u32TotDecimals = 0;

  char      *pcDigits;
  mpfr_exp_t exponentLen;

  // Get the string
  pcDigits = mpfr_get_str(nullptr, &exponentLen, 10, 0, m_mpfr, MPFR_RNDN);
  u32TotDecimals = strlen(pcDigits);

  // Remove the digits before the decimal point
  u32TotDecimals -= exponentLen;

  // Free the resources
  mpfr_free_str(pcDigits);

  return u32TotDecimals;
}
```
Here is the memory after calling **mpfr_get_str**.
Yellow marks the allocated string, and green the debug heap border guards.

![mpfr_get_str](https://user-images.githubusercontent.com/10495173/78983671-edfc5e80-7b24-11ea-83f8-92c6b39c9108.png)

Here is the memory after calling **mpfr_free_str**.
Yellow marks the memory that is overwritten illegally. So, there is underflow damage.

![mpfr_free_str](https://user-images.githubusercontent.com/10495173/78983988-97435480-7b25-11ea-9f94-7ae161d3f012.png)

**Comment by BrianGladman** [comment]
Hi Sander,  Can you please provide more information on this?   First, is this on 32 or 64 bit Windows?  Second, does mpfr.h define mpfr_exp_t as a 32 or 64 bit integer? And, third, is this code that has previously been working and now fails?  I ask because there have been recent changes to enlarge the MPFR exponent range that may have run into the Windows specific issue that long integers on Windows x64 remain as only 32-bit types.   See the discussion here:
https://sympa.inria.fr/sympa/arc/mpfr/2020-03/msg00000.html
Also as an MPFR issue, it is likely that this will need to be reported on the main MPFR site at      https://sympa.inria.fr/sympa

**Comment by SanderBouwhuis** [comment]
1. 64-bit Windows
2. 64-bit build
3. sizeof(mpfr_exp_t) = 4
4. I had the problem today with the mpir+mpfr files as of 20-feb-2020-feb. So, I updated to the files as of 10-apr-2020.

I hadn't noticed it before, but I had an unrelated bug that resulted in a failure of a unit test so I started to investigate. I do have to say that, so far, I've not had any crashes because of it (today).

There may be a possibility that this is normal debug behaviour. But, I'm used to new/delete, and not to malloc/free. So, maybe malloc/free overwrites the debug heap guards in front and behind allocated buffers without problems? It underflows quite a bit, so that surprises me.

**Comment by BrianGladman** [comment]
This is not going to be easy to trace since it is not mpfr that does memory allocation but rather mpir.  In order to track it down I will need to write a bit of test code in C that does exactly what your code does.  I will hence need the exact floating point value that you use to invoke the failure (as a string).

**Comment by SanderBouwhuis** [comment]
Please hold off on any extensive work. I will try to figure out whether malloc has this behaviour. Also, I'll have to check whether the mpfr_get_str function allocates extra memory, which could explain the extra 49 bytes.

**Comment by SanderBouwhuis** [comment]
I followed the flow, and saw this callstack:
mpfr_get_str(nullptr, &exponentlen, 10, 0, m_mpfr, MPFR_RNDN)
mpfr_allocate_func(0 + 310 + 1)
__gmp_default_allocate()
malloc(311)
_malloc_dbg(311, _NORMAL_BLOCK, nullptr, 0)
heap_alloc_dbg(311, _NORMAL_BLOCK, nullptr, 0)
heap_alloc_dbg_internal(311, _NORMAL_BLOCK, nullptr, 0)
_pfnAllocHook(_HOOK_ALLOC, nullptr, 311, _NORMAL_BLOCK, request_number, nullptr, 0)

This in turn changes the size (311) to this:
size += sizeof(_CrtMemBlockHeader) + 1
sizeof(_CrtMemBlockHeader) = 48

In conclusion: the 49 extra underflowing bytes appears to be because the debug version of malloc allocates the extra bytes.

So, case closed I would say.
Brian, if you agree, you can close this issue. Although mpfr_exp_t is still 32-bit on a 64-bit compile, so that may still be an issue.

**Comment by BrianGladman** [comment]
Ok, glad you tracked it down.  The 32-bit mpfr_exp_t on Windows is a pain but the range is still pretty large.
